### DISCUSSION TITLE
Idea: Automatically unsubscribe unknown notification handles

### DISCUSSION THREAD

**Discussion by crishoj** [jisotalo/ads-client#150] [General]
`ads-client` detects incoming notifications for unknown handles, which indicates that previously an ADS subscription was not properly unsubscribed.

In our application code, we detect this by error message `/Notification received with unknown handle \((\d+)\)/` and unsubscribe the handle with a raw ADS command:

```ts
    const handle = Number.parseInt(matches[1])
    const payload = Buffer.alloc(4)
    payload.writeUInt32LE(handle, 0)
    ads.sendAdsCommand({ adsCommand: 7, payload: payload })
```

This works, but could be more convenient, with one or more of these changes:
- [ ] Use a dedicated error class, e.g. `StaleSubscriptionHandleError` with a `handle: number` property.
- [ ] Expose an `unsubscribeRaw(handle: number)` method.
- [ ] Automatically unsubscribe unknown handles (but do emit a warning).

**Comment by jisotalo** [comment]
That sounds like a good idea.

I can't see any reason why deleting unknown ADS notifications received for that port could cause issues. They just use resources.

Perhaps the `unsubscribe()` could also accept numerical handle. That would allow users to do whatever they like:
```js
public async unsubscribe(subscription: ActiveSubscription | number): Promise<void> {
//...
}
```

We could automatically delete all unknown notifications, and write a WARNING: to the console. It could also be a setting such as `deleteUnknownSubscriptions` which defaults to true.
